### PR TITLE
Optimize connection cache eviction logic

### PR DESCRIPTION
**Pull request by pgarg66**
#### Problem
Some comments from #24857 have not been addressed.

#### Summary of Changes
1. Optimize cache eviction logic
2. Some cleanup as suggested in #24857

Fixes #

### CODE REVIEW

File: client/src/connection_cache.rs [hunk 1]
```diff
@@ -241,20 +264,31 @@ fn get_or_add_connection(addr: &SocketAddr) -> GetConnectionResult {
...
+            let mut get_connection_cache_eviction_measure =
+                Measure::start("get_connection_cache_eviction_measure");
+            while map.list_of_peers.len() >= MAX_CONNECTIONS {
                 let mut rng = thread_rng();
                 let n = rng.gen_range(0, MAX_CONNECTIONS);
-                if let Some((nth_addr, _)) = map.map.iter().nth(n) {
-                    let nth_addr = *nth_addr;
-                    map.map.remove(&nth_addr);
-                }
+                let nth_addr = map.list_of_peers.remove(n);
```
Comment (lijunwangs) [inline client/src/connection_cache.rs:276]: remove is O(n), maybe use swap_remove? See https://doc.rust-lang.org/std/vec/struct.Vec.html#method.remove

Comment (pgarg66) [reply client/src/connection_cache.rs:276]: Good point. I'll update the PR.

**Review by lijunwangs — APPROVED** [review]
LGTM

**Comment by lijunwangs** [comment]
We will introduce an Admin RPC call "restartGyserPlugin" which will take a path to the config file of the new plugin that will replace the existing plugin and the config path to the existing plugin (replaced plugin). If the config path is not found or given, it will be treated as simple dynamic plugin load. 

Once the new plugin is loaded, the plugin manager will atomically switch all future notification calls to the new plugin and the  the replaced plugin will be unloaded once all pre-existing calls targeting it return.

**Comment by pgarg66** [comment]
> We will introduce an Admin RPC call "restartGyserPlugin" which will take a path to the config file of the new plugin that will replace the existing plugin and the config path to the existing plugin (replaced plugin). If the config path is not found or given, it will be treated as simple dynamic plugin load. 
> 
> 
> 
> Once the new plugin is loaded, the plugin manager will atomically switch all future notification calls to the new plugin and the  the replaced plugin will be unloaded once all pre-existing calls targeting it return.
> 
> 

Was this comment meant for some other PR?

**Comment by lijunwangs** [comment]
> > We will introduce an Admin RPC call "restartGyserPlugin" which will take a path to the config file of the new plugin that will replace the existing plugin and the config path to the existing plugin (replaced plugin). If the config path is not found or given, it will be treated as simple dynamic plugin load.
> > Once the new plugin is loaded, the plugin manager will atomically switch all future notification calls to the new plugin and the  the replaced plugin will be unloaded once all pre-existing calls targeting it return.
> 
> Was this comment meant for some other PR?

Oh. Sorry. Wrong PR.

**Review by ryleung-solana — APPROVED** [review]

File: client/src/connection_cache.rs [hunk 2]
```diff
@@ -160,7 +172,8 @@ impl ConnectionCacheStats {
 }
 
 struct ConnectionMap {
-    map: BTreeMap<SocketAddr, Connection>,
+    map: HashMap<SocketAddr, Connection>,
+    list_of_peers: Vec<SocketAddr>,
```
Comment (t-nelson) [inline client/src/connection_cache.rs:176]: Seems like a job for [`IndexMap`](https://docs.rs/indexmap/latest/indexmap/index.html) ?

File: client/src/connection_cache.rs [hunk 3]
```diff
@@ -194,32 +208,41 @@ struct GetConnectionResult {
...
+    eviction_timing_ms: u64,
 }
 
 fn get_or_add_connection(addr: &SocketAddr) -> GetConnectionResult {
     let mut get_connection_map_lock_measure = Measure::start("get_connection_map_lock_measure");
     let map = (*CONNECTION_MAP).read().unwrap();
     get_connection_map_lock_measure.stop();
 
-    let mut lock_timing = get_connection_map_lock_measure.as_ms();
+    let mut lock_timing_ms = get_connection_map_lock_measure.as_ms();
```
Comment (t-nelson) [inline client/src/connection_cache.rs:224]: oof. rename made me realize ms is probably a useless resolution for measuring lock contention?

**Review by t-nelson — COMMENTED** [review]
better late than never :sweat_smile:

File: client/src/connection_cache.rs (hunk 3 above)
Comment (pgarg66) [reply client/src/connection_cache.rs:224]: Ideally yes. However, it was taking up to 3000ms earlier. So any value less than 1ms is good indication. It's also easier to plot with other timing values in this function.

File: client/src/connection_cache.rs (hunk 2 above)
Comment (pgarg66) [reply client/src/connection_cache.rs:176]: Yes, that's perfect. I'll update the code.

File: client/src/connection_cache.rs (hunk 3 above)
Comment (t-nelson) [reply client/src/connection_cache.rs:224]: ouch

**Pull request merged**